### PR TITLE
Make address space manager use pagemap for next pointers

### PR DESCRIPTION
**Pull request by mjp41**
The address space manager uses the pagemap entry to form linked lists of
unused address space above MIN_CHUNK_SIZE.  It continues to use
references in the block below that threshold.

In the CHECK_CLIENT mode this makes it hard to corrupt the ASM as only
meta-data uses allocations below MIN_CHUNK_SIZE from the ASM. This
allocations will be protected with guard pages by the backend.

### CODE REVIEW

File: src/backend/address_space_core.h [hunk 1]
```diff
@@ -26,22 +23,19 @@ namespace snmalloc
...
-     * of size 2^n.
-     *
-     * The second entry ranges[n][1] is a pointer to a linked list of blocks
-     * of this size. The final block in the list is not committed, so we commit
-     * on pop for this corner case.
+     * The array indexes based on power of two size.
      *
-     * Invariants
-     *  ranges[n][1] != nullptr => ranges[n][0] != nullptr
+     * The entries for each size form a linked list.  For sizes below,
```
Comment (nwf) [inline src/backend/address_space_core.h:28]:
```suggestion
     * The entries for each size form a linked list.  For sizes below
```

File: src/backend/address_space_core.h [hunk 2]
```diff
@@ -57,45 +51,77 @@ namespace snmalloc
...
     {
-      check_block(base, align_bits);
-      SNMALLOC_ASSERT(align_bits < 64);
-      if (ranges[align_bits][0] == nullptr)
+      if (align_bits >= MIN_CHUNK_BITS)
       {
-        // Prefer first slot if available.
-        ranges[align_bits][0] = base;
+        // The pagemap stores MetaEntrys, abuse the metaslab field to be the
+        // next block in the stack of blocks.
```
Comment (nwf) [inline src/backend/address_space_core.h:72]:
```suggestion
        // next block in the stack of blocks.
        //
        // The pagemap entries here have nullptr (i.e., fake_large_remote) as
        // their remote, and so other accesses to the pagemap (by external_pointer,
        // for example) will not attempt to follow this "Metaslab" pointer.
```

Is that right?

File: src/backend/address_space_core.h [hunk 3]
```diff
@@ -57,45 +51,77 @@ namespace snmalloc
...
+        MetaEntry t(reinterpret_cast<Metaslab*>(next.unsafe_ptr()), nullptr, 0);
+        pagemap.set(address_cast(base), t);
         return;
       }
 
-      if (ranges[align_bits][1] != nullptr)
+      *(reinterpret_cast<CapPtr<void, CBChunk>*>(base.unsafe_ptr())) = next;
```
Comment (nwf) [inline src/backend/address_space_core.h:78]: I fiddled a bit with adding a `struct FreeChunk { CapPtr<FreeChunk, CBChunk> next; };`, which renders this line `base->next = next;`, which is great, and similarly for line 100 below, which just becomes `return base->next;`, but there get to be a lot of casts elsewhere (if you're curious, I've [attached a diff](https://github.com/microsoft/snmalloc/files/6868415/snmalloc-address_space_core-FreeChunk.txt)).  As somewhat sketchy as this line feels, it's probably fine.

Comment (mjp41) [reply src/backend/address_space_core.h:78]: That looks much nicer

File: src/backend/address_space_core.h (hunk 2 above)
Comment (mjp41) [reply src/backend/address_space_core.h:72]: Yes, but also this is memory that is completely controlled by snmalloc, so I don't think anyone should query it with `external_pointer`.  It is not going to be a stack, global, mmap that is done elsewhere in the program.  Though, not entirely sure what the correct full spec of `external_pointer` is for that argument.

**Pull request merged**